internal/pubsub/rabbitpubsub/rabbit.go

```diff
@@ -0,0 +1,179 @@
+
+	// Always use the empty routing key. This driver expects to be used with topic
+	// exchanges, which disregard the routing key.
+	routingKey = ""
+
+	// If the message can't be enqueued, return it to the sender rather than silently dropping it.
+	mandatory = true
+
+	// If there are no waiting consumers, enqueue the message instead of dropping it
+	immediate = true
+)
+
+// OpenTopic returns a *pubsub.Topic corresponding to the named exchange. The
+// exchange must have been previously created (for instance, by using
+// amqp.Channel.ExchangeDeclare). For the model of Go Cloud Pub/Sub to make sense,
+// the exchange should be a fanout exchange, although nothing in this package
+// enforces that.
+//
+// OpenTopic uses the supplied amqp.Connection for all communication. It is the caller's
+// responsibility to establish this connection before calling OpenTopic, and to close
+// it when it is no longer in use.
+//
+// The documentation of the amqp package recommends using separate connections for
+// publishing and subscribing.
+func OpenTopic(conn *amqp.Connection, name string) (*pubsub.Topic, error) {
+	// TODO(jba): support context.Context
+	t, err := newTopic(conn, name)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewTopic(t), nil
+
+}
+
+func newTopic(conn *amqp.Connection, name string) (*topic, error) {
+	t := &topic{
+		exchange: name,
+		conn:     conn,
+	}
+	if err := t.establishChannel(); err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+// establishChannel creates an AMQP channel if necessary. According to the amqp
+// package docs, once an error is returned from the channel, it must be discarded and
+// a new one created.
+//
+// Must be called with the t.mu held.
+func (t *topic) establishChannel() error {
+	// TODO(jba): support context.Context
+	if t.ch != nil {
+		// We already have a channel; nothing to do.
+		return nil
+	}
+	// Create a new channel.
+	ch, err := t.conn.Channel()
+	if err != nil {
+		return err
+	}
+	// Put the channel into a mode where confirmations are delivered for each publish.
+	if err := ch.Confirm(wait); err != nil {
+		return err
+	}
+	t.ch = ch
+	// Get a Go channel which will hold acks from the server. The server
+	// will send an ack for each published message to confirm that it was received.
+	t.pubc = ch.NotifyPublish(make(chan amqp.Confirmation)) // NotifyPublish returns its arg
+	return nil
+}
+
+// SendBatch implements driver.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
+	// It is simplest to allow only one SendBatch at a time. Allowing concurrent
+	// calls to SendBatch would complicate the logic of receiving publish
+	// confirmations and returns. We can go that route if performance warrants it.
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if err := t.establishChannel(); err != nil {
+		return err
+	}
+
+	// Read from the channel established with NotifyPublish.
+	// Do so concurrently or we will be deadlock with the Publish RPC.
```

Comment:
nit: "will be deadlocked" or "will deadlock".

Reply:
done.